### PR TITLE
Add a provision for handling image based bootstrap

### PR DESCRIPTION
**Pull request by yanhua121**
AHV has a limit of 32KB for cloud-init userdata. In Openshift, the ignition can be rather large (a magnitude over the limit). In order to support larger userdata files, we allow mounting the customization as an image. This can be done using the `BootstrapRef` property of the NutanixMachine spec.

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: NutanixMachine
metadata:
  name: machine-name
  namespace: cluster-namespace
spec:
  ...
  bootstrapRef:
    kind: Image
    name: cluster-bootstrap.iso
```

### CODE REVIEW

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.13.0
+    controller-gen.kubebuilder.io/version: (unknown)
```
Comment (deepakm-ntnx) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml:6]: why is this showing unknown

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml [hunk 2]
```diff
@@ -61,20 +66,19 @@ spec:
                         - port
                         type: object
                       failureDomains:
-                        description: failureDomains configures failure domains information
-                          for the Nutanix platform. When set, the failure domains
-                          defined here may be used to spread Machines across prism
-                          element clusters to improve fault tolerance of the cluster.
+                        description: |-
```
Comment (deepakm-ntnx) [inline config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml:69]: are these changes required?

**Comment by deepakm-ntnx** [comment]
pls add unit-test and e2e tests  (tracker if want to do it in separate PR) to verify

File: api/v1beta1/nutanixmachine_types.go [hunk 3]
```diff
@@ -92,6 +92,11 @@ type NutanixMachineSpec struct {
 	// +optional
 	BootstrapRef *corev1.ObjectReference `json:"bootstrapRef,omitempty"`
 
+	// bootstrapImageName is the image name of the bootstrap data.
```
Comment (dkoshkin) [inline api/v1beta1/nutanixmachine_types.go:95]:
```suggestion
	// bootstrapImageName is the name of the image containing additional bootstrap data.
```

File: controllers/nutanixmachine_controller.go [hunk 4]
```diff
@@ -629,6 +604,63 @@ func (r *NutanixMachineReconciler) getOrCreateVM(rctx *nctx.MachineContext) (*nu
...
+			DataSourceReference: &nutanixClientV3.Reference{
+				Kind: utils.StringPtr("image"),
+				UUID: &imgUUID,
+			},
+		}
+		diskList = append(diskList, bootstrapDisk)
+		vmSpec.Resources.DiskList = diskList
+
+		log.Info(fmt.Sprintf("Use the bootstrap image %s to create the VM %s.", bootstrapImgName, vmName))
+	} else {
```
Comment (dkoshkin) [inline controllers/nutanixmachine_controller.go:634]: Is it enough to have one or the other? How is cluster specific data (certs, additional `preKubeadmCommands` and `files`) can be provided here?

File: config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml (hunk 2 above)
Comment (thunderboltsid) [reply config/crd/bases/infrastructure.cluster.x-k8s.io_nutanixclustertemplates.yaml:69]: These seem to be autogenerated. My guess is @yanhua121 is not using devbox for her dependencies and as such the versions of dependencies are different.

@yanhua121 please install devbox and direnv as suggested in the pre-requisites section under https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/main/docs/developer_workflow.md#prerequisites

File: api/v1beta1/nutanixmachine_types.go [hunk 5]
```diff
@@ -92,6 +92,11 @@ type NutanixMachineSpec struct {
...
+	// The image must be available from PC if configured.
+	// +optional
+	BootstrapImageName *string `json:"bootstrapImageName,omitempty"`
```
Comment (thunderboltsid) [inline api/v1beta1/nutanixmachine_types.go:98]: IMO we should add a special field for bootstrap Image. We should instead have a field for additional disks. That is a more generic solution and can be used for mounting disks other than bootstrap disks for other use cases.

**Review by thunderboltsid — CHANGES_REQUESTED** [review]
we need to solve this more generically than we are in this change.

**Review by yanhua121 — COMMENTED** [review]
LGTM. The code changes look good to me. And my local integration test using this branch with the OCP installer passed.

File: controllers/nutanixmachine_controller.go [hunk 6]
```diff
@@ -597,8 +563,14 @@ func (r *NutanixMachineReconciler) getOrCreateVM(rctx *nctx.MachineContext) (*nu
...
 	}
 
-	memorySize := rctx.NutanixMachine.Spec.MemorySize
-	memorySizeMib := GetMibValueOfQuantity(memorySize)
+	diskList, err := getDiskList(rctx)
+	if err != nil {
+		errorMsg := fmt.Errorf("failed to get the disk list to create the VM %s. %v", vmName, err)
+		rctx.SetFailureStatus(capierrors.CreateMachineError, errorMsg)
+		return nil, err
+	}
```
Comment (deepakm-ntnx) [inline controllers/nutanixmachine_controller.go:571]: are we planning to fix all the code codecov warnings? 

controllers/nutanixmachine_controller.go#L566-L570
Added lines #L566 - L570 were not covered by tests

**Review by deepakm-ntnx — COMMENTED** [review]
can we document the yaml input for review as well

File: controllers/nutanixmachine_controller.go (hunk 6 above)
Comment (thunderboltsid) [reply controllers/nutanixmachine_controller.go:571]: Adding code coverage here seems like it will require more refactoring and pulling in a new dependency of Prism Go Client. Given this change needs to make its way into a release-v1.3 branch, I'd like to refrain from making unnecessary changes that would cause problems with cherry-picking. I can offer to make a separate PR for the main branch which will ensure these lines get covered on the main branch.

Comment (deepakm-ntnx) [reply controllers/nutanixmachine_controller.go:571]: ok

**Review by deepakm-ntnx — APPROVED** [review]

**Review by dkoshkin — APPROVED** [review]

**Review by thunderboltsid — APPROVED** [review]
/lgtm
/approve
